com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/SetBreakpointsRequestHandler.java

```diff
@@ -105,17 +109,25 @@ public void initialize(IDebugAdapterContext context) {
         try {
             List<Types.Breakpoint> res = new ArrayList<>();
             IBreakpoint[] toAdds = this.convertClientBreakpointsToDebugger(sourcePath, bpArguments.breakpoints, context);
+            Set<IBreakpoint> processedBreakpoints = new HashSet<>();
             // See the VSCode bug https://github.com/Microsoft/vscode/issues/36471.
             // The source uri sometimes is encoded by VSCode, the debugger will decode it to keep the uri consistent.
             IBreakpoint[] added = manager.setBreakpoints(AdapterUtils.decodeURIComponent(sourcePath), toAdds, bpArguments.sourceModified);
             for (int i = 0; i < bpArguments.breakpoints.length; i++) {
+                // check whether the breakpoint has already been processed, otherwise the same breakpoint might be installed more than once.
+                // if class name is null, then this breakpoint is not valid, we don't need to install it or update hit count for it.
+                if (processedBreakpoints.contains(toAdds[i]) || added[i].className() == null) {
```

Comment:
BreakpointManager checks for duplicate breakpoints. Why do we need to check again?

Reply:
eg: the following code 
    1.  // a
    2. //  b
    3. c++;

if we set all breakpoints for the 3 lines, manager.setBreakpoints will return an 3-element array all with the same breakpoint line # 3, when process the first breakpoint on line 1, actually we are installing breakpoint  on line 3, so the line 2, and line 3 breakpoints will not be processed again.